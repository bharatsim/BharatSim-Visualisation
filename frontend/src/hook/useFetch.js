import { useEffect, useState } from 'react';

import useLoader from './useLoader';
import useDeepCompareMemoize from './useDeepCompareMemoize';

const defaultApiParameters = [];

function useFetch(api, apiParameters = defaultApiParameters, shouldFetch = true) {
  const [responseData, setResponseData] = useState();

  const { loadingState, startLoader, stopLoaderAfterError, stopLoaderAfterSuccess } = useLoader();

  const memoizeValue = useDeepCompareMemoize(apiParameters);

  useEffect(() => {
    if (!shouldFetch) {
      stopLoaderAfterSuccess();
      return;
    }
    startLoader();
    async function fetchApiData() {
      try {
        const resData = await api(...apiParameters);
        setResponseData(resData);
        stopLoaderAfterSuccess();
      } catch (e) {
        stopLoaderAfterError();
      }
    }

    fetchApiData();
  }, memoizeValue);

  return { data: responseData, loadingState: loadingState.state };
}

export default useFetch;
