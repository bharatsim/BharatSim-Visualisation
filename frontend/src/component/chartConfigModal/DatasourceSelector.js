import React from 'react';
import PropTypes from 'prop-types';

import { Box, Typography } from '@material-ui/core';
import Dropdown from '../../uiComponent/Dropdown';

import useFetch from '../../hook/useFetch';
import { api } from '../../utils/api';
import { convertObjectArrayToOptionStructure } from '../../utils/helper';

function DatasourceSelector({ handleDataSourceChange, value, error }) {
  const { data: datasources } = useFetch(api.getDatasources);

  const dataSources = datasources ? datasources.dataSources : [];

  const isDataSourcePresent = dataSources.length > 0;

  return isDataSourcePresent ? (
    <Dropdown
      options={convertObjectArrayToOptionStructure(dataSources, 'name', '_id')}
      onChange={handleDataSourceChange}
      id="dropdown-dataSources"
      label="select data source"
      error={error}
      value={value}
    />
  ) : (
    <Box>
      <Typography>No data source present, upload data source</Typography>
    </Box>
  );
}

DatasourceSelector.defaultProps = {
  error: '',
  value: '',
};

DatasourceSelector.propTypes = {
  handleDataSourceChange: PropTypes.func.isRequired,
  error: PropTypes.string,
  value: PropTypes.string,
};

export default DatasourceSelector;
