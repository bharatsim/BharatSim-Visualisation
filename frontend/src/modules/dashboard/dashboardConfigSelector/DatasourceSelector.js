import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import { Box, Typography } from '@material-ui/core';
import { api } from '../../../utils/api';
import useFetch from '../../../hook/useFetch';
import { convertObjectArrayToOptionStructure } from '../../../utils/helper';
import Dropdown from '../../../uiComponent/Dropdown';
import { projectLayoutContext } from '../../../contexts/projectLayoutContext';

function DatasourceSelector({ handleDataSourceChange, value, error }) {
  const { selectedDashboardMetadata } = useContext(projectLayoutContext);
  const { _id: selectedDashboardId } = selectedDashboardMetadata;
  const { data: datasources } = useFetch(api.getDatasources, [selectedDashboardId]);

  const dataSources = datasources ? datasources.dataSources : [];

  const isDataSourcePresent = dataSources.length > 0;

  return isDataSourcePresent ? (
    <>
      <Box mb={2}>
        <Typography variant="subtitle2"> Data Source </Typography>
      </Box>
      <Dropdown
        options={convertObjectArrayToOptionStructure(dataSources, 'name', '_id')}
        onChange={handleDataSourceChange}
        id="dropdown-dataSources"
        label="select data source"
        error={error}
        value={value}
      />
    </>
  ) : (
    <Box>
      <Typography>No data source present, upload data source</Typography>
    </Box>
  );
}

DatasourceSelector.defaultProps = {
  error: '',
  value: '',
};

DatasourceSelector.propTypes = {
  handleDataSourceChange: PropTypes.func.isRequired,
  error: PropTypes.string,
  value: PropTypes.string,
};

export default DatasourceSelector;
