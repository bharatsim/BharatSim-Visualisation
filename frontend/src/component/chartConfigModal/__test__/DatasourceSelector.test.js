import React from 'react';
import { render, waitFor } from '@testing-library/react';

import useFetch from '../../../hook/useFetch';
import DatasourceSelector from '../DatasourceSelector';

jest.mock('../../../hook/useFetch');

describe('<DatasourceSelector />', () => {
  beforeEach(() => {
    useFetch.mockReturnValue({
      data: {
        dataSources: [
          { name: 'a', _id: 'id_1' },
          { name: 'b', _id: 'id_2' },
        ],
      },
      loadingState: 'SUCCESS',
    });
  });

  it('should match snapshot', () => {
    const { container } = render(
      <DatasourceSelector error="" handleDataSourceChange={jest.fn()} value="" />,
    );

    expect(container).toMatchSnapshot();
  });
  it('should display message of no data source preset if api return empty datasource array', async () => {
    useFetch.mockReturnValue({
      data: { dataSources: [] },
      loadingState: 'SUCCESS',
    });

    const { queryByText } = render(
      <DatasourceSelector error="" handleDataSourceChange={jest.fn()} value="" />,
    );

    await waitFor(() => {
      expect(queryByText(/No data source present, upload data source/i)).not.toBeNull();
    });
  });
});
