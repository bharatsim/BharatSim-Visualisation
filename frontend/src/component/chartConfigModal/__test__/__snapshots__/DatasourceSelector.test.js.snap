// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatasourceSelector /> should match snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-1"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
      data-shrink="false"
      id="dropdown-label"
    >
      select data source
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      data-testid="dropdown-dataSources"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="dropdown-label dropdown-dataSources"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
        id="dropdown-dataSources"
        role="button"
        tabindex="0"
      >
        <span>
          â€‹
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-5"
        >
          <span>
            select data source
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
